File: cores/Rust/compositron/src/core/measurement.rs
use std::collections::{BTreeMap, HashMap};
use std::fmt::{self, Display};

use thiserror::Error;

use crate::importers::slope_n42_importer::{self, import_n42};
use crate::{cdbs::CDBSpectrum, dbs::DBSpectrum};

#[derive(Debug, Error)]
pub enum ImportError {
    #[error("Error during data import")]
    SlopeN42ImportError {
        #[from]
        source: slope_n42_importer::ImportError,
    },

    #[error("Error during data import")]
    CustomImporterError {
        #[source]
        source: Box<dyn std::error::Error + Send + Sync>,
    },
}

#[derive(Debug, Clone, Copy)]
pub enum DataFormat {
    SlopeN42,
    Custom {
        importer: fn (&str) -> Result<Measurement, Box<dyn std::error::Error + Send + Sync>>,
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Shape {
    pub dbs: usize,
    pub cdbs: usize,
}

impl Display for Shape {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f, "Shape(dbs={}, cdbs={})", self.dbs, self.cdbs,
        )
    }
}

pub struct Measurement {
    pub filename: Option<String>,
    pub name: Option<String>,
    pub dbs: BTreeMap<String, DBSpectrum>,
    pub cdbs: BTreeMap<String, CDBSpectrum>,
    pub metadata: HashMap<String, String>,
}

impl Measurement {
    pub fn new() -> Self {
        Measurement {
            filename: None,
            name: None,
            dbs: BTreeMap::new(),
            cdbs: BTreeMap::new(),
            metadata: HashMap::new(),
        }
    }

    pub fn import(
        &mut self, filepath: &str, format: DataFormat
    ) -> Result<(), ImportError> {
        let m = match format {
            DataFormat::SlopeN42 => import_n42(filepath)?,
            DataFormat::Custom { importer } => importer(filepath).map_err(
                |err| ImportError::CustomImporterError { source: err }
            )?
        };

        self.filename = Some(filepath.into());
        self.name = self.filename.clone();
        self.dbs = m.dbs;
        self.cdbs = m.cdbs;
        self.metadata = m.metadata;

        Ok(())
    }

    pub fn shape(&self) -> Shape {
        Shape {
            dbs: self.dbs.len(),
            cdbs: self.cdbs.len(),
        }
    }
}

File: cores/Python/compositron/src/compositron/core/measurement.py
from ..dbs import DBSpectrum
from ..cdbs import CDBSpectrum


class Shape:
    def __init__(self, dbs: int, cdbs: int):
        self.dbs = dbs
        self.cdbs = cdbs

    def __str__(self):
        return repr(self)

    def __repr__(self):
        return f"Shape(dbs={self.dbs}, cdbs={self.cdbs})"


class Measurement:
    def __init__(self, filename: str | None = None):
        self.filename: str | None = filename
        self.name: str | None = filename
        self.dbs: dict[str, DBSpectrum] = {}
        self.cdbs: dict[str, CDBSpectrum] = {}
        self.metadata: dict[str, str] = {}

        if filename is not None:
            from ..importers.SLOPE_n42_importer import import_SLOPE_n42
            m = import_SLOPE_n42(filename)

            self.dbs = m.dbs
            self.cdbs = m.cdbs
            self.metadata = m.metadata

    @property
    def shape(self) -> Shape:
        return Shape(len(self.dbs), len(self.cdbs))
